engine/src/main/java/graphql/nadel/engine/execution/Execution.java

```diff
@@ -163,45 +169,82 @@ private static NadelInstrumentationExecuteOperationParameters buildInstrumentati
 
+        Set<GraphQLOutputType> namespacedGraphqlTypes = overallSchema.getQueryType()
+                .getFieldDefinitions()
+                .stream()
+                .filter(topLevelFieldDef -> !topLevelFieldDef.getDirectives("namespaced").isEmpty())
+                .map(GraphQLFieldDefinition::getType)
+                .collect(Collectors.toSet());
+
+        if (!namespacedGraphqlTypes.isEmpty()) {
+            extractNamespacedFieldInfos(fieldInfoByDefinition, namespacedGraphqlTypes);
+        }
+
         for (Service service : services) {
             List<ObjectTypeDefinition> queryType = service.getDefinitionRegistry().getQueryType();
-            if (queryType != null) {
-                GraphQLObjectType schemaQueryType = overallSchema.getQueryType();
-                for (ObjectTypeDefinition objectTypeDefinition : queryType) {
-                    for (FieldDefinition fieldDefinition : objectTypeDefinition.getFieldDefinitions()) {
-                        GraphQLFieldDefinition graphQLFieldDefinition = schemaQueryType.getFieldDefinition(fieldDefinition.getName());
-                        FieldInfo fieldInfo = new FieldInfo(FieldInfo.FieldKind.TOPLEVEL, service, graphQLFieldDefinition);
-                        fieldInfoByDefinition.put(graphQLFieldDefinition, fieldInfo);
-                    }
+            GraphQLObjectType schemaQueryType = overallSchema.getQueryType();
+            for (ObjectTypeDefinition objectTypeDefinition : queryType) {
+                for (FieldDefinition fieldDefinition : objectTypeDefinition.getFieldDefinitions()) {
+                    GraphQLFieldDefinition graphQLFieldDefinition = schemaQueryType.getFieldDefinition(fieldDefinition.getName());
+                    FieldInfo fieldInfo = new FieldInfo(FieldInfo.FieldKind.TOPLEVEL, service, graphQLFieldDefinition);
+                    fieldInfoByDefinition.put(graphQLFieldDefinition, fieldInfo);
                 }
             }
             List<ObjectTypeDefinition> mutationTypeDefinitions = service.getDefinitionRegistry().getMutationType();
-            if (mutationTypeDefinitions != null) {
-                for (ObjectTypeDefinition mutationTypeDefinition : mutationTypeDefinitions) {
-                    GraphQLObjectType schemaMutationType = overallSchema.getMutationType();
-                    for (FieldDefinition fieldDefinition : mutationTypeDefinition.getFieldDefinitions()) {
-                        GraphQLFieldDefinition graphQLFieldDefinition = schemaMutationType.getFieldDefinition(fieldDefinition.getName());
-                        FieldInfo fieldInfo = new FieldInfo(FieldInfo.FieldKind.TOPLEVEL, service, graphQLFieldDefinition);
-                        fieldInfoByDefinition.put(graphQLFieldDefinition, fieldInfo);
-                    }
+            for (ObjectTypeDefinition mutationTypeDefinition : mutationTypeDefinitions) {
+                GraphQLObjectType schemaMutationType = overallSchema.getMutationType();
```

Comment:
we could move these calls to `overallSchema.getXXXXType` out of the nested loops